blog/2019-12-18-HowToPlanAndOptimizeASecretSanta.adoc

```diff
@@ -0,0 +1,102 @@
+= How to Plan (and Optimize) a Secret Santa
+:page-interpolate: true
+:awestruct-author: cchianel
+:awestruct-layout: blogPostBase
+:awestruct-tags: [useCase]
+
+Many businesses host a "Secret Santa", where each employee gets assigned
+a coworker whom they need to buy a present for. This fosters good relations
+between the employees and brings them closer together. But what about global
+companies, where coworkers may be many miles apart? Ideally, we want
+employees who are further apart to give gifts to each other, since they
+are the ones who probably know the least about each other. Let's optimize
+it with OptaPlanner!
+
+== The Constraints
+
+One of the most obvious constraints is that everyone should get a gift.
+Or phrased a little differently: no one should get multiple gifts.
+
+Here is how one would create this constraint in OptaPlanner using the
+constraint streams API:
+[source,java]
+----
+private Constraint sameReceiverConflict(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .fromUniquePair(SecretSantaAssignment.class,
+                    Joiners.equal(SecretSantaAssignment::getReceiver))
+            .penalize("Same Receiver", HardMediumSoftBigDecimalScore.ONE_HARD);
+}
+----
+
+Next, we need to award assignments where the gifter and the receiver
+are farther apart. Here's how we will do it with constraint streams:
+[source,java]
+----
+private Constraint largerDistanceAward(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .from(SecretSantaAssignment.class)
+            .rewardConfigurableBigDecimal("secretFactor", "Larger Distance Award",
+                              (m) -> BigDecimal.valueOf(Location.calculateDistanceBetween(m.getGifter().getLocation(),
+                                                                       m.getReceiver().getLocation())));
+}
+----
+
+Let's try solving with just these two constraints to see what we get:
+
+image::noPairChecking.png[]
+
+Alice is assigned to Daniella, Daniella is assigned to Alice...
+Bob is assigned to Austin, Austin is assigned to Bob...
+Something about this solution seems odd; it only has assignments
+where the gifter is the receiver's recipient!
+
+== What happened?
+
+This is not odd at all, but rather a consequence of how we defined
+our constraints: if Alice is the person furthest from Daniella, then
+Daniella is most likely the person furthest from Alice. This means our
+"Larger Distance Award" constraint will strongly favour pairs of employees.
+
+Is this a bad thing? Yes it is; imagine if Alice got a $100 gift from
```

Comment:
```suggestion
Is this a bad thing? Yes, it is; imagine if Alice got a $100 gift from
```

Not sure about this.

Reply:
I feel like if we add a comma there, that semi-colon should change to a colon.

Reply:
(I do like it with a pause more)